src/io/dma.rs

```diff
@@ -0,0 +1,209 @@
+//!   registers with all sorts of settings. See the type for more info.
+//!
+//! Note that Source, Destination, and Count are all read-only, while the
+//! Control is read/write. When a DMA unit is _Enabled_ it copies the relevent
+//! Source, Destination, and Count values into its own internal registers (so a
+//! second Enable will reuse the old values). If the DMA _Repeats_ it re-copies
+//! the Count, and also the Destination if
+//! `DMADestAddressControl::IncrementReload` is configured in the Control, but
+//! not the Source.
+//!
+//! When the DMA completes the Enable bit will be cleared from the Control,
+//! unless the Repeat bit is set in the Control, in which case the Enable bit is
+//! left active and the DMA will automatically activate again at the right time
+//! (depending on the Start Timing setting). You have to manually turn off the
+//! correct bit to stop the DMA unit.
+//!
+//! ## Safety
+//!
+//! As you might have noticed by now, utilizing DMA can be very fiddly. It moves
+//! around bytes with no concern for the type system, including the `Clone` and
+//! `Copy` traits that Rust relies on. Use of DMA can be made _somewhat safe_
+//! via wrapper methods (such as those we've provided), but it's fundamentally
+//! an unsafe thing to use.
+//!
+//! ## DMA Can Cause Subtle Bugs
+//!
+//! Since the CPU is halted while DMA is active you can miss out on interrupts
+//! that should have fired. This can cause any number of unintended effects. DMA
+//! is primarily intended for loading large amounts of graphical data from ROM,
+//! or loading sound data at regulated intervals to avoid pops and crackles. It
+//! _can_ be used for general purpose bulk transfers but you are advised to use
+//! restraint.
+
+use super::*;
+
+newtype! {
+  /// Allows you to configure a DMA unit.
+  #[derive(Debug, Copy, Clone, Default, PartialEq, Eq)]
+  DMAControlSetting, u16
+}
+#[allow(missing_docs)]
+impl DMAControlSetting {
+  pub const DEST_ADDR_CONTROL_MASK: u16 = 0b11 << 5;
+  pub fn dest_address_control(self) -> DMADestAddressControl {
+    // TODO: constify
+    match self.0 & Self::DEST_ADDR_CONTROL_MASK {
+      0 => DMADestAddressControl::Increment,
+      1 => DMADestAddressControl::Decrement,
+      2 => DMADestAddressControl::Fixed,
+      3 => DMADestAddressControl::IncrementReload,
+      _ => unsafe { core::hint::unreachable_unchecked() },
+    }
+  }
+  pub const fn with_dest_address_control(self, new_control: DMADestAddressControl) -> Self {
+    Self((self.0 & !Self::DEST_ADDR_CONTROL_MASK) | ((new_control as u16) << 5))
+  }
+
+  pub const SRC_ADDR_CONTROL_MASK: u16 = 0b11 << 7;
+  pub fn src_address_control(self) -> DMASrcAddressControl {
+    // TODO: constify
+    match self.0 & Self::SRC_ADDR_CONTROL_MASK {
+      0 => DMASrcAddressControl::Increment,
+      1 => DMASrcAddressControl::Decrement,
+      2 => DMASrcAddressControl::Fixed,
+      _ => unreachable!(),
```

Comment:
I feel like as a specifically prohibited bit pattern 3 could be given a custom panic message here (pending finding out what actually happens, since you just know if it does anything halfway useful people will want to use it).

Reply:
Well, GBATEK just says "prohibited", so I don't know if you're supposed to know that the bit pattern _never_ happens or what. I'll put it down as a TODO.